design/rest-api-guideline.md

```diff
@@ -0,0 +1,388 @@
+Note that by grouping functionalities together and in conjunction with the versioning schema, we can keep the root path and also the semantics of the API without impact the adaptability for new changes or improvements to the API.
+
+Group names should be in lowercase, a single word name and plural.
+
+#### Versioning
+
+Versioning helps APIs to adapt to changes to their fields or restructure resource representations. In addition to that, versioning can help developers to offer non-stable, technology preview features as well as control the end-of-life of APIs.
+
+The `version` refers to the stability and the state of an API. The version does not necessarily refers to a Keycloak release version number so that it can be mapped to different Keycloak release versions.
+
+Here is an example of a resource path related with the group of functionalities provided by the Account Management REST API:
+
+```
+/apis/accounts/v1/{user_id}
+```
+
+From the example above we can infer that the path refers to a specific `v1` version with the possibility to change the version and: 
+
+* completely change the API
+* do breaking changes to parts of the API
+* remove functionalities or specific parts of the API
+* provide non-stable or technology preview features
+* the different APIs can evolve separately
+
+#### Extensions
+
+To better acommodate and separate the core API from extensions built on top of the API, developers should consider the following path format for extensions:
+
+```
+/apis/{API_GROUP}/extensions/{version}
+```
+
+This separation aims to avoid confusion when consuming stable and official Keycloak APIs so that clients can differentiate if they are using some API provided by a third-party with no support from Keycloak team.
+
+#### Technology Preview
+
+To better acommodate and separate the core API from technology preview features, developers should use a beta version as the API version:
+
+```
+/apis/{API_GROUP}/v1beta1
+```
+
+Technology preview APIs are related with features that have great potential to become fully supported. They provide the latest work in some area of interest and are offered with the intent to gather feedback from the community about new functionalities.
+
+#### Store Resources
+
+A resource store provides operations for a given resource. For example:
+
+```
+/apis/accounts/v1/1234/applications
+```
+
+The example above is representing a resource store path from where we can manage the `applications` for a user account with ID `1234`. Based on this path, clients can perform different operations based on the HTTP verbs or any controller resource defined to the store. Usually this type of resource is associated with CRUD operations for a specific resource.
+
+When a resource fits under this category, the name should be a plural noun.
+
+#### Controller Resources
```

Comment:
Should we not be more prescriptive with regards to controller resources? I'd like to see one way of invoking an operation, not multiple. Either GET or POST, not a choice, etc.

Reply:
I will try to add more info. I tried to give an example using a fictional `/apis/accounts/v1/1234/disable` operation.

I think whether to use GET or POST, depends on the operation. If that is the type of thing you want to have more clarity in this section.

Reply:
I want it as prescriptive as possible, with as little flexibility as possible. I'm fairly certain that we can limit it to POST only. We had some GET based operations in the past, which all as far as I know, have now been converted to POSTs due to bugs, restrictions and security issues (exposing sensitive stuff as query params).

Reply:
Usage of `GET` should consider any security recommendation. But if the actions we are envision for now are all based on `POST`, we can just omit `GET` for now.

Reply:
Based on @vmuzikar comments, we'll use either `POST` or `PUT` for actions. Maybe `PUT` could be used for cases where the action creates additional resource or just updates the resource ? So the client don't need to think about which one to use ?

Reply:
I would just go with POST for all actions. In a way you can read it as "post a new action to be executed" ;)

Reply:
I don't have any preference. I can read PUT in a similar way. I was just wondering whether or not PUT is more correct given that we are operating, and eventually changing, an existing resource.

I can live with POST and just remove references to GET from the document.

Reply:
IMHO `PUT` would be more appropriate for triggering an operation using a controller.

The operation is (usually) a "shortcut" for a resource modification. And `PUT` is intended to be used for modifications.

We can also look at this from a different perspective. Are controller operations idempotent? I believe they should be and therefore `PUT` is the obvious choice (`POST` is not considered to be idempotent).

Reply:
Not always and that is why I mentioned that for me PUT looks more correct. You know what, I think we should leave this open so you could either use PUT or POST. People should respect idempotency (which is already covered in the guideline) when choosing a method.

I was thinking a bit more about GET and despite the security considerations inherent to this method, I think it should also taken into account. For instance, I can think about a GET action that starts some process through redirection.